### ISSUE TITLE
Grouped products's simple products exempt from shopping cart rule.

### ISSUE THREAD

**Issue by apurv-anand** [magento/magento2#6816]
When a simple product (SP) associated to a grouped product is added to the cart, the cart price rules are not getting applied to them. If an individual SP is added to the cart, the rule works fine.

Here is my setup: Grouped Products have associated Simple Products. Both (GP and SPs) of them are in the same category. GPs are visible: Catalog/Search SPs are not visible individually. Magento 2.1.1

My rules are: Condition: If ALL of these conditions are TRUE : If an item is FOUND in the cart with ALL of these conditions true: Category is not one of ...

Action: If ALL of these conditions are TRUE : Category is not one of ...

A SP which is setup individually and is visible gets the rule fine.

I would not like the SPs associated to GPs to be visible in any way.

Steps:
- Setup M 2.1.1
- Create grouped products.
- Create simple products.
- Associate few simple products with GP.
- Set these SPs that are associated to GP as "not visible individually"
- Move them the products to categories. GPs and associated SPs are in same category.
- Create a cart price rule.
- Set a condition to exclude the categories in which GPs and associated SPs are available.
  [Condition: If ALL of these conditions are TRUE : If an item is FOUND in the cart with ALL of these conditions true: Category is not one of ...
  Action: If ALL of these conditions are TRUE : Category is not one of ...
  ]
- Add products to cart, apply the coupon.
- GPs with associated SPs set to invisible wont get the rule even if they should.
- SPs which are individual and set to be visible (catalog / search) would get the rule.

Current: Cart rules do not apply to products set to not visible individually.

Expected: Rules should apply to anything in the cart that fulfills the condition / action. Should not be dependent on their visibility.

**Comment by apurv-anand** [comment]
Need attention on this please.

**Comment by lisaaugustine** [comment]
Our team is also facing exactly the same issue. We need a solution on this as our site is live and we need to use coupons(which we removed for now)..

**Comment by veloraven** [comment]
@apurv-anand thank you for your report.
Could you please clarify expected and actual results?
From your current description it seems there are two problems at once here:
1. Wrong visibility of products.
2. Shopping cart price rule is applied incorrectly.

**Comment by apurv-anand** [comment]
@veloraven 
There is only 1 issue: Shopping cart price rule is applied incorrectly.

Product visibility is fine. The way we have configured the Grouped products is that:
Grouped products are:
Visible to catalog and search.
Simple products associated to these group products are:
Not visible individually

The grouped product are in some category and the associated simple products are also in same category (Ex: Category 1)

A customer can buy these simple products since they will show up in the table below the grouped product (this is expected).

But if a rule is set (Ex: Do not give discount to products in Category 1), the rule will not behave right with these simple products and will still give the discounts. If these simple products are moved to any other visibility ( search or catalog or both) then the rules work fine.

Expected: Rules should apply to not visible individually products too.

**Comment by apurv-anand** [comment]
@veloraven Thanks for marking this as a valid bug report.
Can we take this up on priority? We need a fix for this as soon as possible. Let me know if you can suggest a workaround (even a code change) till then. Making the products visible to search or catalog is not an option for us.

**Comment by apurv-anand** [comment]
Would appreciate a response on prioritizing this issue. Let me know if a hot fix can be provided.

**Comment by dthampy** [comment]
internal jira ticket MAGETWO-60176 is created to track this issue.

**Comment by grazianodimaio** [comment]
Hello,

i have the same issue on a Magento 2.1.0 installation, it's really urgent.

There's an update ?

**Comment by ktruehl** [comment]
Same Problem here. Magento CE 2.1.7.

**Comment by ktruehl** [comment]
I did some debugging and the origin of the problem lies in `Magento\Rule\Model\Condition\Produc\AbstractProduct` in the method `validate()` in the following line
```
return $this->validateAttribute($model->getAvailableInCategories());
```

The method `getAvailableInCategories()` of `Magento\Catalog\Model\ResourceModel\Product` gets the category IDs from the index table `catalog_category_product_index`. However, the index table contains only data for products with visibility `2`, `3` and `4` (i.e. "visibile" products).

So, if the associated product of a grouped product is set to not be visible individually (which makes complete sense for associated products of grouped and configurable products), then the category IDs sales rule condition will not apply to it.

For associated products of configurable products, this is not an issue, since the parent configurable product is part of the quote, i.e. there is a quote item for the parent configurable product, and the category IDs sales rule condition applies to that quote item.

Since we absolutely needed a quick fix for this problem, I created a "hack". It's not pretty, but it gets the job done and it minimizes the danger of side-effects. Here is our patch:
[hack-manually_load_category_ids_for_not_visible_child_products.2017-08-08.diff.zip](https://github.com/magento/magento2/files/1208309/hack-manually_load_category_ids_for_not_visible_child_products.2017-08-08.diff.zip)
